nolabs/domain/models/blast.py

```diff
@@ -0,0 +1,181 @@
+__all__ = [
+    'BlastJob'
+]
+
+import datetime
+from typing import List
+from uuid import UUID
+
+from mongoengine import ReferenceField, EmbeddedDocument, FloatField, EmbeddedDocumentListField, \
+    UUIDField, CASCADE, IntField, StringField
+
+from nolabs.exceptions import NoLabsException, ErrorCodes
+from nolabs.domain.models.common import Job, Protein, JobInputError
+
+
+class Hsp(EmbeddedDocument):
+    num = IntField(required=True)
+    bit_score = FloatField(required=True)
+    score = IntField(required=True)
+    evalue = FloatField(required=True)
+    query_from = IntField(required=True)
+    query_to = IntField(required=True)
+    hit_from = IntField(required=True)
+    hit_to = IntField(required=True)
+    query_frame = IntField(required=True)
+    hit_frame = IntField(required=True)
+    identity = IntField(required=True)
+    positive = IntField(required=True)
+    gaps = IntField(required=True)
+    align_len = IntField(required=True)
+    qseq = StringField(required=True)
+    hseq = StringField(required=True)
+    midline = StringField(required=True)
+
+
+class Hit(EmbeddedDocument):
+    num = IntField(required=True)
+    id = StringField(required=True)
+    definition = StringField(required=True)
+    accession = StringField(required=True)
+    length = IntField(required=True)
+    hsps = EmbeddedDocumentListField(Hsp, required=True)
+
+
+class BlastJobResult(EmbeddedDocument):
+    protein_id = UUIDField(required=True)
+    program = StringField(required=True)
+    database = StringField(required=True)
+    query_id = StringField(required=True)
+    query_def = StringField(required=True)
+    query_len = IntField(required=True)
+    hits = EmbeddedDocumentListField(Hit, required=True)
+
+    @staticmethod
+    def create(protein_id: UUID, program: str, database: str, query_id: str, query_def: str, query_len: int,
+               hits: List[Hit]):
+        return BlastJobResult(
+            protein_id=protein_id,
+            program=program,
+            database=database,
+            query_id=query_id,
+            query_def=query_def,
+            query_len=query_len,
+            hits=hits
+        )
+
+
+class BlastJob(Job):
+    # region Inputs
+
+    protein: Protein = ReferenceField(Protein, reverse_delete_rule=CASCADE, required=True)
+    job_type = StringField(required=True)
+    descriptions = IntField(default=10)
+    alignments = IntField(default=10)
+    hitlist_size = IntField(default=10)
+    expect = FloatField(default=10.0)
+
+    # endregion
+
+    result: List[BlastJobResult] = EmbeddedDocumentListField(BlastJobResult)
+
+    def set_input(self,
+                  protein: Protein,
+                  job_type: str,
+                  descriptions: int = 10,
+                  alignments: int = 10,
+                  hitlist_size: int = 10,
+                  expect: float = 10.0):
+        self.result = []
+
+        self.protein = protein
```

Comment:
Check whether protein has FASTA, because if fasta is missing from the protein domain model for BLAST is broken